src/Streamly/Internal/Unicode/Utf8/Reduce.hs

```diff
@@ -0,0 +1,334 @@
+-- where the predicate returns True for a separator element.  The
+-- resulting components do not contain the separators.  Two adjacent
+-- separators result in an empty component in the output.  eg.
+--
+-- >>> split (=='a') "aabbaca"
+-- ["","","bb","c",""]
+--
+-- >>> split (=='a') ""
+-- [""]
+--
+-- /Time complexity:/ O(n)
+{-# INLINE split #-}
+split :: (Char -> Bool) -> Utf8 -> [Utf8]
+split p t =
+    unsafePerformIO $ Stream.toList $ Stream.splitOn p write (toStream t)
+
+-- | Splits a 'Utf8' into components of length @k@.  The last
+-- element may be shorter than the other chunks, depending on the
+-- length of the input. Examples:
+--
+-- >>> chunksOf 3 "foobarbaz"
+-- ["foo","bar","baz"]
+--
+-- >>> chunksOf 4 "haskell.org"
+-- ["hask","ell.","org"]
+--
+-- /Time complexity:/ O(n)
+{-# INLINE chunksOf #-}
+chunksOf :: Int -> Utf8 -> [Utf8]
+chunksOf k t =
+    unsafePerformIO $ Stream.toList $ Stream.chunksOf k write (toStream t)
+
+-- | Find the first instance of @needle@ (which must be
+-- non-'null') in @haystack@.  The first element of the returned tuple
+-- is the prefix of @haystack@ before @needle@ is matched.  The second
+-- is the remainder of @haystack@, starting with the match.
+--
+-- Examples:
+--
+-- >> breakOn "::" "a::b::c"
+-- ("a","::b::c")
+--
+-- >> breakOn "/" "foobar"
+-- ("foobar","")
+--
+-- Laws:
+--
+-- > append prefix match == haystack
+-- >   where (prefix, match) = breakOn needle haystack
+--
+-- If you need to break a string by a substring repeatedly (e.g. you
+-- want to break on every instance of a substring), use 'breakOnAll'
+-- instead, as it has lower startup overhead.
+--
+-- In (unlikely) bad cases, this function's time complexity degrades
+-- towards /O(n*m)/.
+--
+-- /Time complexity:/ O(n+m)
+{-# INLINE breakOn #-}
+breakOn :: Utf8 -> Utf8 -> (Utf8, Utf8)
+breakOn = undefined
+
+-- XXX Change >> to >>> after implementation
+-- | Similar to 'breakOn', but searches from the end of the
+-- string.
+--
+-- The first element of the returned tuple is the prefix of @haystack@
+-- up to and including the last match of @needle@.  The second is the
+-- remainder of @haystack@, following the match.
+--
+-- >> breakOnEnd "::" "a::b::c"
+-- ("a::b::","c")
+--
+-- /Time complexity:/ O(n+m)
+{-# INLINE breakOnEnd #-}
+breakOnEnd :: Utf8 -> Utf8 -> (Utf8, Utf8)
+breakOnEnd = undefined
+
+--------------------------------------------------------------------------------
+-- Searching
+--------------------------------------------------------------------------------
+
+-- | The 'partition' function takes a predicate and a 'Utf8',
+-- and returns the pair of 'Utf8's with elements which do and do not
+-- satisfy the predicate, respectively; i.e.
+--
+-- > partition p t == (filter p t, filter (not . p) t)
+--
+-- /Time complexity:/ O(n)
+{-# INLINE partition #-}
+partition :: (Char -> Bool) -> Utf8 -> (Utf8, Utf8)
+partition p t = (filter p t, filter (not . p) t)
```

Comment:
Instead of reading it twice, can we use a partition fold to write both the Utf8 segments in one go?

Reply:
My implementation is the bare minimum first thing in my mind kind of a thing. I'll encode the improvement suggesstions. A lot of improvement is required for sure.